File: typescript/server_models.ts
export type ActionOut = FadeActionOut | WaitActionOut | SetActionOut;

export type ActionType = ActionOut | SubSequenceOut | RandomWrapper;

export interface SequenceOut {
    name?: string;
    count?: number;
    endless?: boolean;
    actions: ActionType[];
}

export interface SubSequenceOut {
    type: 'subsequence';
    count: Value;
    actions: ActionType[];
}

export interface FadeActionOut {
    type: 'fade';
    channels: Channels;
    time: Time;
}

export interface WaitActionOut {
    type: 'wait';
    time: Time;
}

export interface SetActionOut {
    type: 'set';
    channels: Channels;
}

export interface RandomWrapper {
    type: 'option';
    options: ActionType[];
}

export type Time = number | TimeWhitespace;

export interface TimeWhitespace {
    type: 'time';
    factor: number;
}

export interface VolumeWhitespace {
    type: 'volume';
    factor: number;
}

export interface RandomValue {
    type: 'random';
    min: number;
    max: number;
}

export type Channels = { [key: string]: ChannelValue };

export type Value = number | RandomValue;
export type ChannelValue = Value | VolumeWhitespace;

export function parseSequence(data: { [key: string]: any }): SequenceOut {
    return {
        count: data.count,
        endless: data.endless,
        actions: data.actions.map((action: any) => recursiveParse(action))
    }
}


function parseValue(data: any): Value {
    if (isRandomValue(data)) {
        return {
            type: 'random',
            min: data.min,
            max: data.max
        }
    }
    return data;
}

function parseChannels(data: { [key: string]: any }): Channels {
    const channels: Channels = {};
    for (const key in data) {
        channels[key] = parseChannelValue(data[key]);
    }
    return channels;
}

function parseChannelValue(data: any): ChannelValue {
    if (isVolumeWhitespace(data)) {
        return {
            type: 'volume',
            factor: data.factor
        }
    }
    return parseValue(data); 
}


function recursiveParse(data: { [key: string]: any }): ActionType {
    if (data.type === 'subsequence') {
        return {
            type: 'subsequence',
            count: parseValue(data.count),
            actions: data.actions.map((action: any) => recursiveParse(action))
        }
    }
    if (data.type === 'option') {
        return {
            type: 'option',
            options: data.options.map((action: any) => recursiveParse(action))
        }
    }
    if (data.type === 'fade') {
        return {
            type: 'fade',
            channels: parseChannels(data.channels),
            time: data.time
        }
    }
    if (data.type === 'wait') {
        return {
            type: 'wait',
            time: data.time
        }
    }
    if (data.type === 'set') {
        return {
            type: 'set',
            channels: parseChannels(data.channels),
        }
    }
    console.log(data);
    throw new Error('Unknown action');
}



export function isSubSequenceOut(action: ActionType): action is SubSequenceOut {
    return action.type === 'subsequence';
}

export function isRandomWrapper(action: ActionType): action is RandomWrapper {
    return action.type === 'option';
}

export function isActionOut(action: ActionType): action is ActionOut {
    return action.type === 'fade' || action.type === 'set' || action.type === 'wait';
}

export function isFadeActionOut(action: ActionType): action is FadeActionOut {
    return action.type === 'fade';
}

export function isSetActionOut(action: ActionType): action is SetActionOut {
    return action.type === 'set';
}

export function isWaitActionOut(action: ActionType): action is WaitActionOut {
    return action.type === 'wait';
}

export function isRandomValue(value: any): value is RandomValue {
    return value !== undefined && typeof value !== 'number' && 'max' in value && 'min' in value;
}

export function isVolumeWhitespace(value: any): value is VolumeWhitespace {
    return value !== undefined && typeof value !== 'number' && 'factor' in value;
}

File: dart/server_models.dart
import 'dart:math';

abstract class Mapable {
  Map<String, dynamic> toMap();
}

abstract class ActionOut extends ActionType {}

abstract class ActionType extends Mapable {}

class SequenceOut extends Mapable {
  int? count;
  bool? endless;
  List<ActionType> actions;

  SequenceOut({this.count, this.endless, required this.actions});

  @override
  Map<String, dynamic> toMap() => {
        'count': count ?? 1,
        'endless': endless,
        'actions': actions.map((action) => action.toMap()).toList(),
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'actions'};
}

class SubSequenceOut extends ActionType {
  String type = 'subsequence';
  Value count;
  List<ActionType> actions;

  SubSequenceOut({required this.count, required this.actions});

  @override
  Map<String, dynamic> toMap() => {
        'type': type,
        'count': count.toMap(),
        'actions': actions.map((action) => action.toMap()).toList(),
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'actions', 'count', 'type'};
}

class FadeActionOut extends ActionOut {
  String type = 'fade';
  Map<String, ChannelValue> channels;
  Time time;

  FadeActionOut({required this.channels, required this.time});

  @override
  Map<String, dynamic> toMap() => {
        'type': type,
        'channels':
            channels.map((channel, value) => MapEntry(channel, value.toMap())),
        'time': time.toMap(),
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'channels', 'time', 'type'};
}

class WaitActionOut extends ActionOut {
  String type = 'wait';
  Time time;

  WaitActionOut({required this.time});

  @override
  Map<String, dynamic> toMap() => {
        'type': type,
        'time': time.toMap(),
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'time', 'type'};
}

class SetActionOut extends ActionOut {
  String type = 'set';
  Map<String, ChannelValue> channels;

  SetActionOut({required this.channels});

  @override
  Map<String, dynamic> toMap() => {
        'type': type,
        'channels':
            channels.map((channel, value) => MapEntry(channel, value.toMap())),
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'channels', 'type'};
}

class RandomWrapper extends ActionType {
  String type = 'option';
  List<ActionType> options;

  RandomWrapper({required this.options});

  @override
  Map<String, dynamic> toMap() => {
        'type': type,
        'options': options.map((option) => option.toMap()).toList(),
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'options', 'type'};
}

abstract class Time {
  const Time();

  dynamic toMap();
}

class NumTime extends Time {
  final num value;

  NumTime(this.value);

  const NumTime.ms75() : value = 75;

  const NumTime.ms200() : value = 200;

  const NumTime.ms1000() : value = 1000;

  @override
  toMap() => value;

  static bool fits(dynamic data) => data is num;
}

class TimeWhitespace extends Time {
  String type = 'time';
  num factor;

  TimeWhitespace({required this.factor});

  @override
  toMap() => {
        'type': type,
        'factor': factor,
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'type', 'factor'};
}

class VolumeWhitespace extends ChannelValue {
  String type = 'volume';
  double factor;

  VolumeWhitespace({required this.factor});

  @override
  toMap() => {
        'type': type,
        'factor': factor,
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'type', 'factor'};
}

class RandomValue extends Value {
  String type = 'random';
  int min;
  int max;

  RandomValue({required this.min, required this.max});

  int get value => min + Random().nextInt(max - min);

  @override
  toMap() => {
        'type': type,
        'min': min,
        'max': max,
      };

  static bool fits(Set<String> keys) => keys.containsAll(requiredKeys);

  static Set<String> get requiredKeys => {'type', 'min', 'max'};
}

typedef Channels = Map<String, ChannelValue>;

abstract class Value extends ChannelValue {
  const Value();
}

class NumValue extends Value {
  final int value;

  NumValue(this.value);

  const NumValue.zero() : value = 0;

  const NumValue.full() : value = 255;

  @override
  toMap() => value;

  static bool fits(dynamic data) => data is num;
}

abstract class ChannelValue {
  const ChannelValue();
  dynamic toMap();
}

bool isActionOut(Map<String, dynamic> map) {
  final keys = map.keys.toSet();
  if (SetActionOut.fits(keys)) return true;
  if (FadeActionOut.fits(keys)) return true;
  if (WaitActionOut.fits(keys)) return true;
  return false;
}

bool isSubSequenceOut(Map<String, dynamic> map) {
  final keys = map.keys.toSet();
  if (SubSequenceOut.fits(keys)) return true;
  return false;
}

bool isActionType(Map<String, dynamic> map) {
  final keys = map.keys.toSet();
  if (isActionOut(map)) return true;
  if (SubSequenceOut.fits(keys)) return true;
  if (SequenceOut.fits(keys)) return true;
  if (RandomWrapper.fits(keys)) return true;
  return false;
}

bool isChannelValue(dynamic data) {
  if (data is Map) {
    final keys = data.keys.map((e) => e.toString()).toSet();
    if (RandomValue.fits(keys)) return true;
    if (VolumeWhitespace.fits(keys)) return true;
  }
  if (NumValue.fits(data)) return true;
  return false;
}

bool isTime(dynamic data) {
  if (data is Map) {
    final keys = data.keys.map((e) => e.toString()).toSet();
    if (TimeWhitespace.fits(keys)) return true;
  }
  if (NumTime.fits(data)) return true;
  return false;
}

bool isValue(dynamic data) {
  if (data is Map) {
    final keys = data.keys.map((e) => e.toString()).toSet();
    if (RandomValue.fits(keys)) return true;
  }
  if (NumValue.fits(data)) return true;
  return false;
}

SequenceOut parseSequenceOut(Map<String, dynamic> map) {
  return SequenceOut(
    endless: map['endless'],
    count: map['count'],
    actions: List.from(map['actions'])
        .map<ActionType>((action) => parseActions(action))
        .toList(),
  );
}

ActionType parseActions(dynamic data) {
  final map = Map<String, dynamic>.from(data);
  final type = map['type'];
  if (type == 'option') {
    return RandomWrapper(
      options: List.from(map['options'])
          .map<ActionType>((option) => parseActions(option))
          .toList(),
    );
  }
  if (type == 'subsequence') {
    return SubSequenceOut(
      count: parseValue(data['count']),
      actions: List.from(map['actions'])
          .map<ActionType>((action) => parseActions(action))
          .toList(),
    );
  }
  if (type == 'fade') {
    return FadeActionOut(
      channels: parseChannels(map['channels']),
      time: parseTime(map['time']),
    );
  }
  if (type == 'wait') {
    return WaitActionOut(
      time: parseTime(map['time']),
    );
  }
  if (type == 'set') {
    return SetActionOut(
      channels: parseChannels(map['channels']),
    );
  }
  throw Exception('Could not parse action: $type');
}

Channels parseChannels(Map<String, dynamic> map) {
  return map.map(
    (channel, value) => MapEntry(channel, parseChannelValue(value)),
  );
}

Time parseTime(dynamic data) {
  if (data is Map) {
    final keys = data.keys.map((e) => e.toString()).toSet();
    if (TimeWhitespace.fits(keys)) {
      return TimeWhitespace(
        factor: data['factor'],
      );
    }
  }
  if (NumTime.fits(data)) {
    return NumTime(data);
  }
  throw Exception('Could not parse time');
}

Value parseValue(dynamic data) {
  if (data is Map) {
    final keys = data.keys.map((e) => e.toString()).toSet();
    if (RandomValue.fits(keys)) {
      return RandomValue(
        min: data['min'],
        max: data['max'],
      );
    }
  }
  if (NumValue.fits(data)) {
    return NumValue(data);
  }
  throw Exception('Could not parse value: $data');
}

ChannelValue parseChannelValue(dynamic data) {
  if (data is Map) {
    final keys = data.keys.map((e) => e.toString()).toSet();
    if (RandomValue.fits(keys)) {
      return RandomValue(
        min: data['min'],
        max: data['max'],
      );
    }
    if (VolumeWhitespace.fits(keys)) {
      return VolumeWhitespace(
        factor: data['factor'],
      );
    }
  }
  if (NumValue.fits(data)) {
    return NumValue(data);
  }
  throw Exception('Could not parse channel value');
}
